### PR TITLE
Add callsInPlace to ReceiveChannel.consume

### LINKED ISSUE #941: Add callsInPlace exactly once contract for ReceiveChannel.consume
If you agree this should be added, I can make a PR.

### PR DESCRIPTION
**Pull request by qwwdfsad**
Fixes #941

### CODE REVIEW

File: kotlinx-coroutines-core/common/test/BuilderContractsTest.kt [hunk 1]
```diff
@@ -44,9 +45,18 @@ class BuilderContractsTest : TestBase() {
...
+        i.consume {
+            ch = 321
+        }
+        consume(ch)
     }
 
     private fun consume(a: Int) {
-        a.hashCode() // BE codegen verification
+        assertNotEquals(0, a)
+        assertEquals(a.hashCode(), a)
```
[the remark this answers is not in the document]
Comment (qwwdfsad) [reply kotlinx-coroutines-core/common/test/BuilderContractsTest.kt:60]: Nothing to fancy, just dummy verification of the value.
If the value wasn't set (e.g. contract is misplaced), two things could happen: either bytecode verifier would trigger a `VerificationError` (most likely) or the value would have its default value.

I'm testing both here: value is not zero and some method is invoked on it (and `int.hashCode() == int` by its contract).

Do you think it worth a comment?

[the remark this answers is not in the document]
Comment (dkhalanskyjb) [reply kotlinx-coroutines-core/common/test/BuilderContractsTest.kt:60]: It is clear why these properties must be true, but it's not evident at all why they may be violated, so I definitely think that some form of documentation must be added here.
> `// BE codegen verification`

did provide at least some context–it showed that it was the compiler's output that was checked–but without it, calls to `consume()` look like weird noops. So, at least preserving that comment would be of huge help.

Though maybe it's standard practice to add such checks when working with contracts?

[the remark this answers is not in the document]
Comment (qwwdfsad) [reply kotlinx-coroutines-core/common/test/BuilderContractsTest.kt:64]: Left a comment

File: kotlinx-coroutines-core/common/test/BuilderContractsTest.kt [hunk 2]
```diff
@@ -56,6 +56,10 @@ class BuilderContractsTest : TestBase() {
     }
 
     private fun consume(a: Int) {
+        /*
+         * Verify the value is actually set correctly
+         * (non-zero, VerificationError is not triggered can be read)
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-core/common/test/BuilderContractsTest.kt:61]: There's probably a comma missing here?

**Review by dkhalanskyjb — APPROVED** [review]
Looks good.

File: kotlinx-coroutines-core/common/test/BuilderContractsTest.kt (hunk 2 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-core/common/test/BuilderContractsTest.kt]: thanks, fixed

**Pull request merged**